src/inc/conpty-universal.h

```diff
@@ -30,18 +30,161 @@
 
 const unsigned int PTY_SIGNAL_RESIZE_WINDOW = 8u;
 
-HRESULT CreateConPty(const std::wstring& cmdline,       // _In_
-                     const unsigned short w,            // _In_
-                     const unsigned short h,            // _In_
-                     HANDLE* const hInput,              // _Out_
-                     HANDLE* const hOutput,             // _Out_
-                     HANDLE* const hSignal,             // _Out_
-                     PROCESS_INFORMATION* const piPty); // _Out_
+// A case-insensitive wide-character map is used to store environment variables
+// due to documented requirements:
+//
+//      "All strings in the environment block must be sorted alphabetically by name.
+//      The sort is case-insensitive, Unicode order, without regard to locale.
+//      Because the equal sign is a separator, it must not be used in the name of
+//      an environment variable."
+//      https://docs.microsoft.com/en-us/windows/desktop/ProcThread/changing-environment-variables
+//
+struct WStringCaseInsensitiveCompare
+{
+    bool operator()(const std::wstring& lhs, const std::wstring& rhs) const
+    {
+        return (::_wcsicmp(lhs.c_str(), rhs.c_str()) < 0);
+    }
+};
+
+using EnvironmentVariableMapW = std::map<std::wstring, std::wstring, WStringCaseInsensitiveCompare>;
+
+HRESULT CreateConPty(const std::wstring& cmdline,                       // _In_
+                     const unsigned short w,                            // _In_
+                     const unsigned short h,                            // _In_
+                     HANDLE* const hInput,                              // _Out_
+                     HANDLE* const hOutput,                             // _Out_
+                     HANDLE* const hSignal,                             // _Out_
+                     PROCESS_INFORMATION* const piPty,                  // _Out_
+                     const EnvironmentVariableMapW& extraEnvVars = {}); // _In_
 
 bool SignalResizeWindow(const HANDLE hSignal,
                         const unsigned short w,
                         const unsigned short h);
 
+HRESULT UpdateEnvironmentMapW(EnvironmentVariableMapW& map);

```

Comment:
should these functions be marked `static ` (file-scoped)?

Reply:
This is a header file, so that would mean a copy of the function in every one that calls it, so I didn't bother since I wouldn't want that effect anyway.

If the concern is to document to callers the fact that this is really private mechanism, then the best I can think of doing is placing this into a ::detail namespace as recommended in the ISO C++ guidelines.

I actually wrote these functions generically with the assumption that environment manipulation might be useful elsewhere.

With that in mind, maybe the environment functions I added are best moved to something like utils.cpp?  (Ah, utils, the faithful dumping ground of all projects).

Reply:
Oh yea, that's fine then. I think if someone else really needs it, we'll let them move it to utils themselves :P